### ISSUE TITLE
Autoresize Helm

### ISSUE THREAD

**Issue by ReneFroger** [emacs-helm/helm#791]
Hello! 

Thanks for maintaining and extending this nice plugin! I'm trying Emacs (with Evil) and Helm. One of the things that I already miss, is the autoresizing the window, depending on the results that you get. 

For example, I made a video about Unite which have auto-resizing enabled: https://www.youtube.com/watch?v=fwqhBSxhGU0

I studied the documentation of Helm, and googled various tutorials about Helm like the Helm tutorial from tuhdo. But I couldn't find it: is there any way to enable autoresize in Helm? 

Or have IDO that only? 

Thanks in advance for your reply.

**Comment by tuhdo** [comment]
It is that way because Helm has an action menu to perform various actions. What if you narrow to a single candidate? Helm buffer will be minimized to a single line and it would be annoying when you switch back to the action menu and the buffer is enlarged again, or the action menu is displayed with just a single line making it difficult to navigate the menu.

Helm buffer also has to adapt with small windows. For example, if you create 4 smaller windows (an Emacs window is not OS window; it's the same as a pane in tmux) in your Emacs frame, then you will not want Helm buffer to be any smaller otherwise it would be really difficult to use or unusable. I don't recall in your Unite video you work with multiple smaller panes. You should try to see how Unite behaves in such situation.

**Comment by ReneFroger** [comment]
Thanks for your reply. I wouldn't find it annoying if the buffer is enlarged again, when I switch back. As you can see in the video, Unite have the same behaviour. It's an option in Unite actually, you can enable the auto-resizing. 

So I would like to see this as an option in Emacs. I believe it's possible, when you look how golden-ratio package in Emacs works. Then this feature could communicate the lines to the window-manager. 

I tried multiple smaller panes in Vim (8 small panes, actually). And maybe it will surprise you, but Unite behaves well with these panes. Unite have a some sort of  'overlay' (my English is not good), which acts as top on the viewport, instead nesting it. So you don't have this issue.

**Comment by ReneFroger** [comment]
When I tried it in Helm, it seems it's using the same technique as Unite. The windows will not change, when you click on the bar of Helm, and move it up and down. So perhaps it's possible to add variable for autoresizing, depending on the lines of results? My thank will be eternally.

**Comment by thierryvolpiatto** [comment]
ReneFroger <notifications@github.com> writes:

> When I tried it in Helm, it seems it's using the same technique as
> Unite. The windows will not change, when you click on the bar of Helm,
> and move it up and down. So perhaps it's possible to add variable for
> autoresizing, depending on the lines of results? My thank will be
> eternally.

Use C-} and C-{ to re-size manually helm window.
for example, when you want to browse images in a bigger size from
helm-find-files, hit C-t repeatedly to swap windows as needed, then
re-size with C-} and then hit C-<down> to see this/next image.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by thierryvolpiatto** [comment]
ReneFroger <notifications@github.com> writes:

> Thanks for your reply. I wouldn't find it annoying if the buffer is enlarged again, when I switch back. As you can see in the video, Unite have the same behaviour.
> It's an option in Unite actually, you can enable the auto-resizing.

You can try something like this:

(add-hook 'helm-after-update-hook (lambda () (with-helm-window (fit-window-to-buffer))))

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by thierryvolpiatto** [comment]
Tu Do <notifications@github.com> writes:

> It is that way because Helm has an action menu to perform various actions. What if you narrow to a single candidate? Helm buffer will be minimized to a single line
> and it would be annoying when you switch back to the action menu and the buffer is enlarged again, or the action menu is displayed with just a single line making it
> difficult to navigate the menu.
>
> Helm buffer also has to adapt with small windows. For example, if you create 4 smaller windows (an Emacs window is not OS window; it's the same as a pane in tmux) in
> your Emacs frame, then you will not want Helm buffer to be any smaller otherwise it would be really difficult to use or unusable. I don't recall in your Unite video
> you work with multiple smaller panes. You should try to see how Unite behaves in such situation.

This issue can be fixed with helm-always-two-windows == t.
I always use this.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by thierryvolpiatto** [comment]
@tuhdo Could you make a small video or gif with those settings:

```elisp
(add-hook 'helm-after-update-hook (lambda () (with-helm-window (fit-window-to-buffer))))
(setq helm-always-two-windows  t)
```

**Comment by ReneFroger** [comment]
>> You can try something like this:

>> (add-hook 'helm-after-update-hook (lambda () (with-helm-window (fit-window-to-buffer))))

That's great! I really couldn't find it in the tutorials/manual. Thanks for your help, it's truly appreciated.

**Comment by tuhdo** [comment]
@thierryvolpiatto Sure. Should it be added to the doc a section named "Autoresize" as well?

**Comment by tuhdo** [comment]
@thierryvolpiatto I think this is a good chance to introduce a feature like `autoresize-mode`, since it is already available. So, instead of using this code:

```elisp
(add-hook 'helm-after-update-hook (lambda () (with-helm-window (fit-window-to-buffer))))
(setq helm-always-two-windows  t)
```

we could simply use this:

```elisp
(autoresize-mode 1)
```

Is it ok?

**Comment by tuhdo** [comment]
We can add support `helm-autoresizing-mode`, with a maximum height to avoid Helm buffer overlaps most of Emacs frame that causes this issue: https://github.com/emacs-helm/helm/issues/795 

```elisp
(defcustom helm-autoresize-max-height 20)

(defun helm-autoresize-hook ()
  (with-helm-window (fit-window-to-buffer (get-buffer-window) helm-autoresize-max-height)))

(add-hook 'helm-after-update-hook 'helm-autoresize-hook)
```

The above hook is added only when `helm-autoresize-mode` is activated, and is removed when deactivated.

**Comment by thierryvolpiatto** [comment]
Tu Do <notifications@github.com> writes:

> We can add support helm-autoresizing-mode, with a maximum height to
> avoid Helm buffer overlaps most of Emacs frame that causes this issue:
> #795

Good, I will add this ASAP unless you beat me with a PR ;-)

Thanks.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by tuhdo** [comment]
@ReneFroger  You should leave the issue open until the feature is available.

**Comment by yveszoundi** [comment]
What about just using/recommending packages such as [popwin](https://github.com/m2ym/popwin-el) in this case?

**Comment by tuhdo** [comment]
Not everyone uses popwin and it does not solve this issue anyway.  What the OP needs id a way to automatically resize Helm buffer according to the number of candidates. You should try the autoresize code abovd to see the effect.

**Comment by thierryvolpiatto** [comment]
So now please remove the hook in your config and use only:

```elisp
(helm-auto-resize-mode 1)
```

**Comment by tuhdo** [comment]
Awesome. I will make a gif and update the documentation later.

**Comment by jixiuf** [comment]
I use https://github.com/roman/golden-ratio.el
I could 
```
(add-to-list 'golden-ratio-exclude-modes "calendar-mode") ; this work
```

```
(add-to-list 'golden-ratio-exclude-modes "helm-mode") ; this does not work
(helm-autoresize-mode 1)
```
so it seems there is not a helm-mode or helm-buffer-mode 
```
 MR Name                                                    Size Mode             Filename/Process
 -- ----                                                    ---- ----             ----------------
[ Helm ]
 m  *helm M-x*                                               116 Helm             
 m  *helm for files*                                        6.9k Helm             
 m  *helm lsgit*                                             125 Helm             
 m  *Helm file completions*                                  828 Helm             
 m  *helm grep*                                              404 Helm             

```

**Comment by tuhdo** [comment]
You can inhibit golden-ratio in Helm buffer like this:

```elisp
 (require 'golden-ratio)

  (defun pl/helm-alive-p ()
    (if (boundp 'helm-alive-p)
        (symbol-value 'helm-alive-p)))

  (add-to-list 'golden-ratio-inhibit-functions 'pl/helm-alive-p)
```

**Comment by jixiuf** [comment]
thanks.

**Comment by ReneFroger** [comment]
I found out that I had a harder time to read the contents from Helm with autoresize than with Unite (a Vim plugin with same functionality like Helm). 

Now I know why. 

In Unite, when you type, the result is matched in the first line above your input. So if the results of Everything fills your buffer, you can still focus on the first results. 

But in Helm with autorezise, my most relevant results (the recentf/buffer files) are moving around, because the Locate results below keeps them moving. 

So I wonder if there is a way to sort them in the order of:

locate
files-from-current directory
recentf
buffers

This would make much more easier to not get distracted by Helm Locate. I wonder if that's possible?

**Comment by tuhdo** [comment]
Yes. `helm-for-files` uses `helm-for-files-preferred-list` as it source, and `helm-for-files-preferred-list` is composed of:

```elisp
'(helm-source-buffers-list
  helm-source-recentf
  helm-source-bookmarks
  helm-source-file-cache
  helm-source-files-in-current-dir
  helm-source-locate)
```

The list contains smaller sources. So, to reoder the result, you simply reorder the sources. In your case, you want to move locate at the top, simply do this:

```lisp
(setq helm-for-files-preferred-list '(helm-source-locate
                                      helm-source-buffers-list
                                      helm-source-recentf
                                      helm-source-bookmarks
                                      helm-source-file-cache
                                      helm-source-files-in-current-dir))
```

**Comment by thierryvolpiatto** [comment]
ReneFroger <notifications@github.com> writes:

> So I wonder if there is a way to sort them in the order of:
>
> locate
> files-from-current directory
> recentf
> buffers

No there is no way to do this, because locate is an asynchronous source,
i.e while helm-locate is computing its results, the other sources are
directly computed and sent in helm-buffer before the locate result
arrive.
This among other nuisances to have *-locate in helm-for-files.

My workflow with locate is start to find the file with helm-find-files,
<ok I don't know where to search> and from there hit C-x C-f which run
*-locate to continue the search deeper.

From helm-find-files, you can also hit C-c / ("find"), or C-x C-d
(browse project).

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by ReneFroger** [comment]
@thierryvolpiatto why is it not possible? From Tuhdo's answer it seems to be possible to change the order of the Helm sources? 

@tuhdo Thanks for your reply. 

As you can propably noticed in the video of [Unite](https://www.youtube.com/watch?v=fwqhBSxhGU0), I can keep my eyes focused while Unite is autoresizing, because the most relevant results are above the text input, not on the top of the autoresing Unite window. 

Or it would be hard to not be distracted then. 

So I wonder if it's possible to get the relevant results in the Helm on the line above text input, (the last line in Helm buffer), instead on first line of helm buffer? 
It would make much more easier to keep the focus. :smiley:

**Comment by tuhdo** [comment]
@ReneFroger  It seems like reordering like that has no effect on async source.

Well, there was an issue for it: https://github.com/emacs-helm/helm/issues/362. As Le Wang suggested, you should not look at the bottom but look at the top. That's why Helm did not have the autoresize feature in the first place. The best results are always at the top, which is closer to the current buffer you are working in, so you don't have to move your eyes between current buffer and the minibuffer at the bottom frequently. It is better that way because if you use Helm commands frequently, you don't constantly move your eyes.

That's said, `helm-autoresize-mode` still has its benefit: 

- It is less distracting when you set max/min not too far apart from each other.
- Or set Helm window to smaller size if you set max equal to min, without relying on another package like popwin thaht is not really compatible with Helm.

**Comment by thierryvolpiatto** [comment]
ReneFroger <notifications@github.com> writes:

> @thierryvolpiatto why is it not possible? From Tuhdo's answer it seems
> to be possible to change the order of the Helm sources?

Even if you change the order of sources the async source (locate)
results will be the last inserted in helm-buffer due to the nature of
async sources.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by ReneFroger** [comment]
@tuhdo I think it would be more benefit because you are looking on the input line/command line to see what you were typing, and then get the most relevant lines directly above the input line/command line, instead on the top. 

The top is constantly moving, max-equal will not help a much, because if you get only less results than the max-equal lines, the top line will moving from his sport, anyway. 

@thierryvolpiatto I understand that, but that's not the problem here, I guess?

**Comment by tuhdo** [comment]
@ReneFroger What I meant was that the buffer your eyes will have to travel less the the buffer you are editing.

@thierryvolpiatto That's being said, `helm-autoresize-mode` is most useful when the sort in reverse feature is available. Do you think it would be possible if after all candidates are sorted and inserted into a Helm buffer, then we move one by one from top to bottom to achieve the reverse effect, then move point to `(point-min)` then `(flush-line "^$")`?

**Comment by ReneFroger** [comment]
@tuhdo now I get it. Thanks for your explanation. Then it amazes me that the Everything in Unite (smilar to Locate in Helm) is an asynchronous process. 

@thierryvolpiatto have already mentioned it's has something to do with the process handling in Emacs. I would like to give up some little speed, to get Helm his results at down, above input minibuffer.
As far as it's now, the autoresizing feature of Helm is not usable. It makes you very hard to use when the relevant lines are at top, so I will not using it. But thanks for all your replies! :+1:

**Comment by tuhdo** [comment]
@ReneFroger Are you sure it handles the way Helm does? For example, when you grep a large source tree, you will have to wait a bit before seeing results, correct? In Helm, as soon as the first result is returned, it is handled to Helm immediately for displaying. And you can get something like 5000 grep hits in a Helm buffer for scrolling up/down or save in another buffer for mass replace. In the video, I didn't see Unite does thing like scrolling the list of candidates. I guess the way Unite does is that as soon as it fills the visible area, it terminates the external process immediately until you you start typing again, no?

**Comment by ReneFroger** [comment]
I know that Unite uses Vimproc in order to asynchronously deliver the results, but maybe @Shougo can answer that about Vimproc. For the rest, frankly, I don't know how the internals work. I have a little experience with VimL and Elisp, but not enough to answer your reply. Sorry. :)

**Comment by tuhdo** [comment]
And VimProc has the core written in C: https://github.com/Shougo/vimproc.vim/tree/master/autoload

**Comment by PythonNut** [comment]
@tuhdo I am a Unite user as well. As far as I can tell, `Unite` calls out to external processes just like Helm. The default `files_rec/async` uses `find`. You do need to wait for relevant results to be produced. Minor note that Unite does no caching, so you must wait on _every_ search.

@ReneFroger 
Some helm processes are asynchronous. `helm-source-locate` is one of them. The thing about `locate` is that is doesn't take much time to run, but it produces a massive blast of output by default. (Remember that locate itself is a cached operation that runs in practically constant time). `locate` rarely runs for more that a second (except to hang around and blast its output), so making it not only asynchronous but asynchronous with asynchronously streamed candidates makes little sense.

`helm-scource-locate` is not comparable to `files_rec/async`. If you want a helm source that duplicates `files_rec/async` use `helm-source-findutils`.

`find`, on the other hand, is not cached, and it's generally pretty slow, so making it stream candidates is a good idea. In fact, that is what helm does.

**Comment by tuhdo** [comment]
I did a quick and terrible hack just to see how sort in reverse look in Helm. Note that this is just a demo; you won't be able to select anything other than seeing the results:

```elisp
(defun helm--sort-in-reverse ()
  (let (kill-ring)
    (goto-char (point-min))
    (kill-line)
    (cl-loop for i from 1 to (count-lines (point-min) (point-max))
             do (kill-line))
    (cl-loop for l in kill-ring
             do (insert l))))

(add-hook 'helm-after-update-hook 'helm--sort-in-reverse)
```

After you're done with it, remove it with:

```elisp
(remove-hook 'helm-after-update-hook 'helm--sort-in-reverse)
```

It seems to work for async source as well, but slow when it has a heap of results. Set `helm-candidate-number-limit` to something smaller than the default like 50 or less, it's as fast as normal.

**Comment by thierryvolpiatto** [comment]
ReneFroger <notifications@github.com> writes:

> @thierryvolpiatto have already mentioned it's has something to do with
> the process handling in Emacs. I would like to give up some little
> speed, to get Helm his results at down, above input minibuffer.  As
> far as it's now, the autoresizing feature of Helm is not usable. It
> makes you very hard to use when the relevant lines are at top, so I
> will not using it.

As I said before, but apparently you skipped it, make helm display its
results on the upper window, like this the top of your helm window
(where the first results are) will not move.

See helm-split-window-default-side, what you need is 'above.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by tuhdo** [comment]
Maybe this will help a bit: make Helm window always at the bottom, and small enough to lessen the travel distance between top match and minibuffer using [popwin](https://github.com/m2ym/popwin-el):

Initial setup:

```elisp
(require 'popwin)
(popwin-mode 1)

(global-set-key (kbd "C-c b") popwin:keymap) ; bind to a key of your choice
```

Then add this (add more to the list as you want):

```elisp
(setq popwin:special-display-config
      '(
        ("^\*helm.+\*$" :regexp t :height 12 :position bottom)
        ))
````

The above code means that `popwin` displays any Helm buffer with the height 12 (lines) and display at the bottom. It will always open Helm window at the bottom, not inside any other window and it won't temporary hide other windows you are working with. It reduces the travel distance between top matches and the minibuffer, the smaller the Helm window is.

In the long run, I think you should adapt with the Helm way. Sure, Unite has that nice feature to reverse the output, I admit. But I don't think it's much of an issue if you can touch type (I assume). I always look at the Helm buffer when I type and look at it most of the time after finishing typing. Only when I need to adjust my input patterns, I look at the minibuffer to move the cursor to appropriate place for entering new pattern or delete unnecessary one. Is there any major problem for you with such usage?

And I already posted a demo code to see that Helm can display the same thing as Unite fine. But then, there are higher priority issues at the moment.

**Comment by ReneFroger** [comment]
@tuhdo Thanks for your thoughts. I always do touchtyping, but the only issue here is that before I press Enter, I would like to see that the right file is selected. The travel distance between the minibuffer and top is not relevant here, but it's a nice bonus. :smiley:
It's just when Helm is auto resizing, it makes me hard to check if I really selected the right file. So that's why autoresize is not useful for me now. 

But thanks anyway, your help is really appreciated. 

By the way, you mentioned Grizzl and it's not maintained anymore. I noticed in my feed that a [well known Emacs'er is beginning to take over the maintaining of Grizzl](https://twitter.com/bbatsov/status/555262136348585984). 

For the rest, it will makes me to not use autoresize feature of Helm. But it's just a tiny detail for me. It's not so important indeed and this issue should be considered as closed. :-)

**Comment by tuhdo** [comment]
@ReneFroger Yes, it was not maintained since 2013 until today, Batsov decided to take over it. `grizzl` might be suitable for people who just want a simple interface without extra utilities. I tried `grizzl` on a fairly large candidate list and it hangs when I inserted a few characters. That's why I like Helm so much. Even for a list with 80k candidates, it narrows fine, even with fuzzy matching, which is amazing, even though it took a while to retrieve candidates. Note that I have an SSD and ran that test on Linux, so it was possible. The number would be lower for normal hard drive and less powerful machines, on different OS.
